### PR TITLE
IBX-37: Implemented data storage contracts

### PR DESCRIPTION
**Pull request by ciastektk**
| Question                                  | Answer
| ---------------------------------------- | ------------------
| **JIRA issue**                          | [IBX-187](https://issues.ibexa.co/browse/IBX-187)
| **Type**                                   | feature
| **Target Ibexa DXP version** | `v4.0.0`
| **BC breaks**                          | no
| **Doc needed**                       | yes

This PR provides added contracts for data storage. 
In next steps content based data storage will be implemented.
Partners and external clients should add own implementation of data storage which will allow to export no content data by personalization-client and use in personalization module.

#### Checklist:
- [x] Provided PR description.
- [ ] Tested the solution manually.
- [ ] Provided automated test coverage.
- [x] Checked that target branch is set correctly (master for features, the oldest supported for bugs).
- [x] Ran PHP CS Fixer for new PHP code (use `$ composer fix-cs`).
- [x] Asked for a review (ping `@ezsystems/php-dev-team`).

### CODE REVIEW

**Review by adamwojs — APPROVED** [review]

File: src/contracts/Storage/CriteriaInterface.php [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+    /**
+     * @return array<string>
+     */
+    public function getLanguages(): array;
+
+    /**
+     * @return ?array<string>
+     */
+    public function getItemIds(): ?array;
```
Comment (Steveb-p) [inline src/contracts/Storage/CriteriaInterface.php:26]: In what cases do we expect this method to return `null`?

If it's supposed to be a case for something being unsupported, then it should either be it's own interface or throw exceptions?

**Review by Steveb-p — APPROVED** [review]

File: src/contracts/Storage/CriteriaInterface.php (hunk 1 above)
Comment (ciastektk) [reply src/contracts/Storage/CriteriaInterface.php:26]: My point is to add possibility to narrow down the results by given identifiers for specific item types and languages. I thought by default it should be null but if someone will want to fetch only specific items then they need to return array of item ids.

File: composer.json [hunk 2]
```diff
@@ -34,7 +34,8 @@
     "autoload": {
         "psr-4": {
             "EzSystems\\EzRecommendationClientBundle\\": "src/bundle/",
-            "EzSystems\\EzRecommendationClient\\": "src/lib/"
+            "EzSystems\\EzRecommendationClient\\": "src/lib/",
+            "Ibexa\\Contracts\\PersonalizationClient\\": "src/contracts/"
```
Comment (alongosz) [inline composer.json:38]: I would rather see this as:
```suggestion
            "Ibexa\\Contracts\\Personalization\\Client\\": "src/contracts/"
```
@adamwojs @Steveb-p WDYT?

**Review by alongosz — CHANGES_REQUESTED** [review]

File: composer.json (hunk 2 above)
Comment (Steveb-p) [reply composer.json:38]: I'm ok with either. The second can introduce class name collisions if there would be a package with `Ibexa\Contracts\Personalization` classes that adds `Client` sub-namespace, but I think we can manage that.

:+1:

File: src/contracts/Storage/CriteriaInterface.php (hunk 1 above)
Comment (Steveb-p) [reply src/contracts/Storage/CriteriaInterface.php:26]: Then it needs a comment in phpdoc explaining that returning `null` from this method has a specific meaning / significance.

Another option is creating a "hasser", which would enable us to drop the nullability of the return. In theory that's a "cleaner" solution, but I'm not sure if it's worth it?

Comment (alongosz) [reply src/contracts/Storage/CriteriaInterface.php:26]: > My point is to add possibility to narrow down the results by given identifiers for specific item types and languages. I thought by default it should be null but if someone will want to fetch only specific items then they need to return array of item ids.

Could you provide an example of a code that would utilize this feature?

File: composer.json (hunk 2 above)
Comment (ciastektk) [reply composer.json:38]: This repository will be moved to ezplatform-personalization (ibexa/personalization) so maybe better idea is to add just `"Ibexa\\Contracts\\Personalization\\": "src/contracts/"`?

File: src/contracts/Storage/CriteriaInterface.php (hunk 1 above)
Comment (ciastektk) [reply src/contracts/Storage/CriteriaInterface.php:26]: After long thought I decided to drop this method. If someone will need this kind of feature then must implement it by it self.

**Comment by ciastektk** [comment]
I will fix this (https://github.com/ezsystems/ezrecommendation-client/pull/51/commits/467c574d94222f58e9e6966a32afe5b84a768aef) in next step when I will focus on the refactoring

**Review by alongosz — APPROVED** [review]

**Review by adamwojs — APPROVED** [review]

**Review by Steveb-p — APPROVED** [review]

**Pull request merged**